modules/pricer/src/main/java/com/opengamma/strata/pricer/impl/tree/SingleBarrierKnockoutFunction.java

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (C) 2016 - present by OpenGamma Inc. and the OpenGamma group of companies
+ * 
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.pricer.impl.tree;
+
+import java.util.Arrays;
+
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.collect.array.DoubleArray;
+import com.opengamma.strata.collect.array.DoubleMatrix;
+import com.opengamma.strata.product.fx.BarrierType;
+
+/**
+ * Single barrier knock-out option function.
+ */
+abstract class SingleBarrierKnockoutFunction implements OptionFunction {
+
+  /**
+   * Obtains the barrier level for the time layer specified by {@code step}.
+   * 
+   * @param step  the step
+   * @return the barrier level
+   */
+  public abstract double getBarrierLevel(int step);
+
+  /**
+   * Obtains the sign.  
+   * <p>
+   * The sign is +1 for call and -1 for put.
+   * 
+   * @return the sign
+   */
+  public abstract double getSign();
+
+  /**
+   * Obtains the barrier type. 
+   * 
+   * @return the barrier type
+   */
+  public abstract BarrierType getBarrierType();
+
+  /**
+   * Obtains the rebate for the time layer specified by {@code step}.
+   * 
+   * @param step  the step
+   * @return the rebate
+   */
+  public abstract double getRebate(int step);
+
+  @Override
+  public DoubleArray getPayoffAtExpiryTrinomial(DoubleArray stateValue) {
+
+    int nNodes = stateValue.size();
+    double[] values = new double[nNodes];
+    double rebate = getRebate(getNumberOfSteps());
+    double barrierLevel = getBarrierLevel(getNumberOfSteps());
+    boolean isDown = getBarrierType().isDown();
+    Arrays.fill(values, rebate);
+    int index = getLowerBoundIndex(stateValue, barrierLevel);
+    ArgChecker.isTrue(index > -1 && index < nNodes - 1, "barrier is covered by tree");
+    int iMin = isDown ? index + 1 : 0;
+    int iMmax = !isDown ? index + 1 : nNodes;
+    for (int i = iMin; i < iMmax; ++i) {
+      values[i] = Math.max(getSign() * (stateValue.get(i) - getStrike()), 0d);
+    }
+    double bd = barrierLevel - stateValue.get(index);
+    double ub = stateValue.get(index + 1) - barrierLevel;
+    double ud = stateValue.get(index + 1) - stateValue.get(index);
+    if (isDown) {
```

Comment:
Worth a comment on why the value is modified here. If there was only the expiry date valuation, this "if-else" would not be there.